zkvm/src/ops.rs

```diff
@@ -399,4 +400,55 @@ impl Program {
         self.0.push(Instruction::Cloak(m, n));
         self
     }
+
+    /// Takes predicate and closure to add choose operations for
+    /// predicate tree traversal.
+    pub fn choose_predicate<F, T>(
+        &mut self,
+        pred: Predicate,
+        choose_fn: F,
+    ) -> Result<&mut Program, VMError>
+    where
+        F: FnOnce(PredicateTree) -> Result<T, VMError>,
+    {
+        choose_fn(PredicateTree {
+            prog: self,
+            pred: pred,
+        })?;
+        Ok(self)
+    }
+}
+
+/// Adds data and instructions to traverse a predicate tree.
+pub struct PredicateTree<'a> {
+    prog: &'a mut Program,
+    pred: Predicate,
+}
+
+impl<'a> PredicateTree<'a> {
+    /// Left Predicate branch
+    pub fn left(self) -> Result<Self, VMError> {
+        let (l, r) = self.pred.to_disjunction()?;
+        let prog = self.prog;
+        prog.push(l.clone()).push(r).left();
```

Comment:
Since we know that we are going to consume these predicates in the `left`/`right` instructions, we don't need to clone the entire predicate tree (`l.clone()`): we can convert to an opaque predicate. Moving `r` into push comes for free, but we can make it opaque too for symmetry.

Let's add `Predicate::to_opaque(&self) -> Predicate` method that returns `Predicate::Opaque(self.to_point())` and do this:

```suggestion
        prog.push(l.to_opaque()).push(r.to_opaque()).left();
```

Same thing for `right()`.

WDYT?

Reply:
that sounds good to me :)